qemu_system.go

```diff
@@ -0,0 +1,432 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+
+	govmmQemu "github.com/intel/govmm/qemu"
+)
+
+type qemuArch interface {
+	// enableNestingChecks nesting checks will be honoured
+	enableNestingChecks()
+
+	// disableNestingChecks nesting checks will be ignored
+	disableNestingChecks()
+
+	// machine returns the machine type
+	machine() (govmmQemu.Machine, error)
+
+	// qemuPath returns the path to the QEMU binary
+	qemuPath() (string, error)
+
+	// kernelParameters returns the kernel parameters
+	// if debug is true then kernel debug parameters are included
+	kernelParameters(debug bool) []Param
+
+	//capabilities returns the capabilities supported by QEMU
+	capabilities() capabilities
+
+	// bridges returns the number bridges for the machine type
+	bridges(number uint32) []Bridge
+
+	// cpuTopology returns the CPU topology for the given amount of vcpus
+	cpuTopology(vcpus uint32) govmmQemu.SMP
+
+	// cpuModel returns the CPU model for the machine type
+	cpuModel() string
+
+	// memoryTopology returns the memory topology using the given amount of memoryMb and hostMemoryMb
+	memoryTopology(memoryMb, hostMemoryMb uint64) govmmQemu.Memory
+
+	// append9PVolumes appends volumes to devices
+	append9PVolumes(devices []govmmQemu.Device, volumes []Volume) []govmmQemu.Device
+
+	// appendConsole appends a console to devices
+	appendConsole(devices []govmmQemu.Device, path string) []govmmQemu.Device
+
+	// appendImage appends an image to devices
+	appendImage(devices []govmmQemu.Device, path string) ([]govmmQemu.Device, error)
+
+	// appendBridges appends bridges to devices
+	appendBridges(devices []govmmQemu.Device, bridges []Bridge) []govmmQemu.Device
+
+	// append9PVolume appends a 9P volume to devices
+	append9PVolume(devices []govmmQemu.Device, volume Volume) []govmmQemu.Device
+
+	// appendSocket appends a socket to devices
+	appendSocket(devices []govmmQemu.Device, socket Socket) []govmmQemu.Device
+
+	// appendNetwork appends a endpoint device to devices
+	appendNetwork(devices []govmmQemu.Device, endpoint Endpoint) []govmmQemu.Device
+
+	// appendBlockDevice appends a block drive to devices
+	appendBlockDevice(devices []govmmQemu.Device, drive Drive) []govmmQemu.Device
+
+	// appendVhostUserDevice appends a vhost user device to devices
+	appendVhostUserDevice(devices []govmmQemu.Device, vhostUserDevice VhostUserDevice) []govmmQemu.Device
+
+	// appendVFIODevice appends a VFIO device to devices
+	appendVFIODevice(devices []govmmQemu.Device, vfioDevice VFIODevice) []govmmQemu.Device
+}
+
+type qemuSystem struct {
```

Comment:
I think it'd be more appropriate to rename `qemuSystem` into `qemuArchBase` in order to understand that no one is supposed to use this implementation. It is meant to be embedded into a "child" implementation.